### PR TITLE
[ot] Do not drop data on PTY write if it is connected

### PR DESCRIPTION
**Pull request by pamaury**
The current code in char-pty.c is a bit odd: is uses poll to determine when it can read from a device, but it does not poll for write. Instead, it uses a timer to regularly determine whether the PTY is connected. We can always determine the connection status of a PTY by polling the HUP flag (up for disconnected, down for connected). This commit refactors the code to get rid of the timer and simply poll with no delay on every write to see if it connected.

A better solution would be to use an I/O watch with a callback.

**Context:** we want to use a PTY to connect the UART to opentitantool. However with the current code, the connection status might take up to a second to be updated so if we want to avoid dropping data from early boot, we need to wait a whole second doing nothing on startup. This will prevent this delay.

### CODE REVIEW

File: chardev/char-pty.c [hunk 1]
```diff
@@ -49,64 +48,35 @@ DECLARE_INSTANCE_CHECKER(PtyChardev, PTY_CHARDEV,
...
     PtyChardev *s = PTY_CHARDEV(chr);
     GPollFD pfd;
     int rc;
     QIOChannelFile *fioc = QIO_CHANNEL_FILE(s->ioc);
 
     pfd.fd = fioc->fd;
-    pfd.events = G_IO_OUT;
+    /* If a PTY is connected, the HUP flag is cleared so
+     * we can poll the connection status this way. */
+    pfd.events = G_IO_HUP;
```
Comment (jwnrt) [inline chardev/char-pty.c:61]: For opentitantool we have exclusive access to the PTY so this isn't an issue, however do you know what happens with HUP when there are two readers and only one of them disconnects? Can you even have two readers on a PTY?

File: chardev/char-pty.c [hunk 2]
```diff
@@ -49,64 +48,35 @@ DECLARE_INSTANCE_CHECKER(PtyChardev, PTY_CHARDEV,
...
-    } else {
-        pty_chr_state(chr, 1);
-    }
+    pty_chr_state(chr, !(pfd.revents & G_IO_HUP));
+}
+
+static void pty_chr_update_read_handler(Chardev *chr)
+{
+    char_pty_poll_connection_status(chr);
 }
```
Comment (jwnrt) [inline chardev/char-pty.c:72]: What does this function wrapper do?

**Review by jwnrt — APPROVED** [review]
Thanks for finding and fixing this @pamaury

I think I understand the change: remove the timer code and do the poll on writes as well as the existing reads.

Is this patch suitable for upstreaming to QEMU? I'm guessing you don't think so based on:

> A better solution would be to use an I/O watch with a callback.

**Comment by rivos-eblot** [comment]
I think you need to discuss this with QEMU upstream. PTY and chardev support is a delicate piece of SW in QEMU, and not breaking stuff might be difficult. You also need to check it works on other Unix flavors (*BSD, macOS, ...) - just trying not opening a can of worms :-)

BTW can't you use sockets to connect the UART?

**Comment by jwnrt** [comment]
> BTW can't you use sockets to connect the UART?

Yes, it just means implementing in-band software flow control manually rather than relying on existing code for serial ports that connect to TTYs.

We've had code mostly working using sockets before, and may go back to using them if this doesn't work out

File: chardev/char-pty.c (hunk 1 above)
Comment (pamaury) [reply chardev/char-pty.c:61]: I tried on my machine and I could not get a second terminal to connect (it said that the device is locked).

**Comment by rivos-eblot** [comment]
> > BTW can't you use sockets to connect the UART?

> Yes, it just means implementing flow control manually rather than relying on existing code for serial ports that connect to TTYs.

Do  you need flow control at all in this case?

File: chardev/char-pty.c (hunk 2 above)
Comment (pamaury) [reply chardev/char-pty.c:72]: `pty_chr_update_read_handler` is using by QEMU to update the reading side of the device. I am sure a QEMU export can fill the blanks, but the bottom line is that this function was already there and polled the FD. I just factored out the code to `char_pty_poll_connection_status`.

**Comment by jwnrt** [comment]
> Do you need flow control at all in this case?

I'm not sure, I was fine running the OpenTitan UART tests that I tried, but opentitantool has implemented flow control support for the other "transports" (verilator and FPGAs) so I was thinking there might be some tests that rely on it.

Maybe we can go ahead using sockets without flow control and see if anything breaks.

**Comment by pamaury** [comment]
@jwnrt thinks it's simpler to reuse the PTY code that we have for real hardware since it's supported by QEMU.

I don't know if we can upstream that change, ideally yes but as you said it's probably quite tedious to check it on every single platform.

**Comment by rivos-eblot** [comment]
Note that sockets in QEMU are raw sockets so if you need to use out-of-bad data, you will need to provide your own chardev backend. It seems quite some work - not sure it is worth it.

**Comment by rivos-eblot** [comment]
> @jwnrt thinks it's simpler to reuse the PTY code that we have for real hardware since it's supported by QEMU.

I would tend to think it is much simpler to change OT tools than attempting to change anything upstream :-) (personal note: I would never touch this piece of code myself, too many risks of unexpected regressions)

**Comment by jwnrt** [comment]
Thanks for the guidance @rivos-eblot, I'm going to try shuffle opentitantool around so we can reuse the software flow control code over a socket as well as the PTY

**Comment by pamaury** [comment]
By some incredible luck, someome posted a very similar patch to the qemu devel mailing list a few days ago:
https://lore.kernel.org/qemu-devel/20230816210743.1319018-1-thuth@redhat.com/
I have sent a message saying that we are interested in seeing this change happen.

**Comment by jwnrt** [comment]
I've updated my opentitantool PR to use files as the UART connection so it can share the verilator code.

If that patch lands in upstream QEMU, maybe we can cherry pick it in and switch to using the serial directly, but for now the file-based connection seems to work fine.

**Comment by rivos-eblot** [comment]
> I've updated my opentitantool PR to use files as the UART connection so it can share the verilator code.

Is this PR https://github.com/lowRISC/opentitan/pull/19471 ?

**Comment by jwnrt** [comment]
> Is this PR [lowRISC/opentitan#19471](https://github.com/lowRISC/opentitan/pull/19471) ?

Yes, that's right.

That PR is based on / waiting on https://github.com/lowRISC/opentitan/pull/19420 which is for having Bazel pass the QEMU flags. We're just missing that flag to ignore the incorrect flashes.

**Comment by jwnrt** [comment]
> By some incredible luck, someome posted a very similar patch to the qemu devel mailing list a few days ago: https://lore.kernel.org/qemu-devel/20230816210743.1319018-1-thuth@redhat.com/ I have sent a message saying that we are interested in seeing this change happen.

That fix landed upstream ~~in (I think) 8.2.0~~, so maybe we can close this PR ~~and wait until the next rebase on QEMU~~: https://github.com/qemu/qemu/commit/4f7689f0817a717d18cc8aca298990760f27a89b

**Comment by jwnrt** [comment]
Apologies, it actually landed pre-8.1.2 so we have the fix in our tree now. I'll close this PR.

**Pull request closed without merging**